File: LicenseIOSAppByNamedUser_ObjC/LicenseIOSAppByNamedUser_ObjC/LicenseiOSAppViewController.m
//
//  ViewController.m
//  LicenseIOSAppByNamedUser_ObjC
//
//  Created by maklMac on 2/28/17.
//  Copyright © 2017 esrichina.com. All rights reserved.
//

#import "LicenseiOSAppViewController.h"
#import "LoginViewController.h"
#import <ArcGIS/ArcGIS.h>

@interface LicenseiOSAppViewController ()

@property (weak, nonatomic) IBOutlet AGSMapView *mapView;

@property (nonatomic, strong) AGSMap *map;

@property (weak, nonatomic) IBOutlet UIVisualEffectView *saveAsBlurView;


@property (weak, nonatomic) IBOutlet UIToolbar *savingToolbar;

@property(strong,nonatomic) LoginViewController* loginVC;

@property(strong,nonatomic)AGSPortal* portal;



@end

@implementation LicenseiOSAppViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    self.map = [[AGSMap alloc] initWithBasemap:[AGSBasemap topographicBasemap]];

    self.mapView.map = self.map;
    
    self.saveAsBlurView.hidden = true;
}

-(void)loginPortalwithName:(NSString *)name key:(NSString *)key{
    
    self.portal = [AGSPortal portalWithURL:[NSURL URLWithString:@"https://www.arcgis.com"] loginRequired:true];
    
    self.portal.credential = [AGSCredential credentialWithUser:name password:key];
    
    [self.portal loadWithCompletion:^(NSError * error){
        
        if (error != nil) {
            
            NSLog(@"Login portal failed: @%@",error);
            
        }else{
            
            NSLog(@"portal login succeed!");
            AGSLicenseInfo* licenseInfo = self.portal.portalInfo.licenseInfo;
            
            @try{
                NSError* err;
                [AGSArcGISRuntimeEnvironment setLicenseInfo:licenseInfo error:&err];
                
            }
            @catch(NSError* error) {
                
                NSLog(@"error:@%@",error);
            }
            
            NSDictionary *licenseDictionary;
            
            @try{
                NSError* err;
                licenseDictionary = [licenseInfo toJSON:&err];
                
            }
            @catch(NSError* error) {
                
                NSLog(@"LicenseInfo not available, error:@%@",error);
            }

            AGSKeychainItem *keychainItem = [[AGSKeychainItem alloc] initWithIdentifier:@"maklesri.LicenseIOSAppByNamedUser" accessGroup:nil acrossDevices:false];
            
            [keychainItem writeObjectToKeychain:licenseDictionary completion:^(NSError * wirteError){
                if(wirteError){
                    NSLog(@"Error writing to the keychain...@%@",wirteError);
                }
            }];
        }
    }];
    
}

//MARK: - Actions
- (IBAction)ActivateLicense:(id)sender {
    
    [self toggleLoginView];
}


//MARK: - hide/show input screen
-(void)toggleLoginView{
    
    self.saveAsBlurView.hidden = !self.saveAsBlurView.hidden;
}

#pragma mark - Navigation
// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
    
    
    self.loginVC = [segue destinationViewController];
    
    self.loginVC.delegate = self;
}

 //MARK: - LoginVCDelegate
-(void)loginViewController:(LoginViewController*)LoginViewController didInitiateLoginWithName:(NSString *)name key:(NSString*)key{
    
    [self loginPortalwithName:name key:key];
    
    [self toggleLoginView];
    
    
}

-(void)loginViewControllerDidCancel:(LoginViewController*)LoginViewController{
    [self toggleLoginView];
    
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: LicenseIOSAppByNamedUser_Swift/LicenseIOSAppByNamedUser_Swift/LicenseiOSAppViewController.swift
//
//  AppDelegate.swift
//  LicenseIOSAppByNamedUser_Swift
//
//  Created by maklMac on 2/14/17.
//  Copyright © 2017 esrichina.com. All rights reserved.
//

import UIKit
import ArcGIS


class LicenseiOSAppViewController: UIViewController,LoginVCDelegate {
    
    
    @IBOutlet private weak var mapView:AGSMapView!

    @IBOutlet private weak var saveAsBlurView:UIVisualEffectView!
    @IBOutlet private weak var savingToolbar:UIToolbar!
    
    
    
    private var loginVC:LoginViewController!
    
    private var portal:AGSPortal!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
    
        let map = AGSMap(basemap: AGSBasemap.imagery())
        
        self.mapView.map = map
        

    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    
    
  
    
    private func toggleLoginView() {
        
        self.saveAsBlurView.isHidden = !self.saveAsBlurView.isHidden
        
    }
    
    //MARK: - Actions
    @IBAction func ActivateLicense(_ sender: Any) {
        
        self.toggleLoginView()
        
    }
    
    func LoginPortal(withname name:String,key:String) {
        
        self.portal = AGSPortal(url: URL(string: "https://www.arcgis.com")!, loginRequired: true)
        
        self.portal.credential = AGSCredential(user: name, password: key)
        
        self.portal.load { (error) in
            if let error = error {
                
                print("Login portal failded：\(error)")
                
            }
            else {
                
                //get title etc
                print("portal login succeed!")
                
                //get the users license info from the portal info
                let licenseInfo = self.portal.portalInfo?.licenseInfo
                
                
                //set the license using the Named Users licenseInfo
                do {
                    
                    let result = try AGSArcGISRuntimeEnvironment.setLicenseInfo(licenseInfo!)
                    
                }
                catch let error as NSError {
                    print("error: \(error.localizedDescription)")
                }
                    
                
                
                //The app is now licensed
                
                // Save the license information so that the app can be started and licensed offline
                // In this example the licenseDictionary is stored in keychain
                
                var licenseDictionary: NSDictionary?
                
                do {licenseDictionary = try licenseInfo?.toJSON() as! NSDictionary?
                } catch {
                    print("LicenseInfo not available")
                }
                
                
                let keychainItem = AGSKeychainItem(identifier: "maklesri.LicenseIOSAppByNamedUser", accessGroup: nil, acrossDevices: false)
                
                keychainItem.writeObject(toKeychain: licenseDictionary!, completion: { (writeError) in
                    if let error = writeError {
                        print("Error writing to the Keychain... \(error)")
                    }
                })
                
            }
            
        }
    }

    
    
    @IBAction private func cancelAction() {
        self.view.endEditing(true)
        self.toggleLoginView()
    }
    
    //MARK: - Navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        
        if segue.identifier == "LoginEmbedSegue" {
            
            print("正在初始化LoginViewController")
            
            self.loginVC = segue.destination as! LoginViewController
            
            
            self.loginVC.delegate = self
            
        }
        
    }
    
    
    //MARK: - LoginVCDelegate
    
    func loginViewController(_ loginViewController:LoginViewController, didInitiateLoginWithName name:String, key:String){
        
        self.LoginPortal(withname:name,key:key)
        
        self.toggleLoginView()
    }
    
    func loginViewControllerDidCancel(_ saveAsViewController:LoginViewController){
        
        self.toggleLoginView()
    }
    
    
    

}
